swaywm/sway
# Implement fullscreen

This implements `fullscreen enable`, `fullscreen disable`, `fullscreen toggle`, and allows views to request fullscreen.

This is my first significant PR for sway and I've had to learn a lot about how everything works, so please scrutinise this carefully.

Note that sway 0.15 allows the command `fullscreen` without any arguments which acts as toggle, but this is different to the i3 user guide so I've implemented it as described above. As a result, if you're porting config from sway 0.15 you may need to update your command to be `fullscreen toggle`.

wl_shell seems a bit lacking with its fullscreen ability, and from what I understand wl_shell is deprecated anyway, so I've not implemented fullscreen for that. If someone could confirm then that'd be appreciated.

There is one known issue, and that is that popups don't show when the parent view is fullscreen, but I'd like to either resolve that in a separate PR or leave it for someone else to look into.

Tests:

* Run fullscreen enable, fullscreen disable, fullscreen toggle on an xdg_v6 view
* Run fullscreen enable, fullscreen disable, fullscreen toggle on an xwayland view
* Attempt to fullscreen a container - should be ignored 
* Fullscreen a view, then kill it
* Open two views on one workspace, fullscreen one, then attempt to focus the other - focus should remain on fullscreen view
* Open two views on different workspaces, fullscreen one, then attempt to focus the other and back again - workspaces should focus correctly
* Fullscreen view, then move view to another workspace - view should be unfullscreened
* Make an xwayland view request fullscreen, eg. launch firefox and press F11
* Make an xdg_v6 view request fullscreen, eg. launch `mpv --opengl-backend=wayland` and press f
